htlcswitch/link.go

```diff
@@ -1637,8 +1637,21 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 				OnionSHA256: msg.ShaOnionBlob,
 			}
 		default:
-			log.Errorf("Unknown failure code: %v", msg.FailureCode)
-			failure = &lnwire.FailTemporaryChannelFailure{}
+			log.Warnf("Unexpected failure code received in "+
+				"UpdateFailMailformedHTLC: %v", msg.FailureCode)
+
+			// We don't just pass back the error we received from
+			// our successor. Otherwise we might report a failure
+			// that penalizes us more than needed. If the onion that
+			// we forwarded was correct, the node should have been
+			// able to send back its own failure. The node did not
+			// send back its own failure, so we assume there was a
+			// problem with the onion and report that back. We reuse
+			// the invalid onion key failure because there is no
+			// specific error for this case.
+			failure = &lnwire.FailInvalidOnionKey{
```

Comment:
IIRC, in the past, we decided to send the temp chan failure as we didn't want to give away any further information detailing that we rejected the packet due to it being a replay. However, I think this is ok as well since it doesn't give away any additional information.

Reply:
One thing that I was wondering is why a replay always leads to failing back the htlc. Shouldn't the receiving node just ignore the duplicate package so that the htlc can live on?

Reply:
change to invalid key makes sense to me.

> Shouldn't the receiving node just ignore the duplicate package so that the htlc can live on?

If the packet is replayed it will result in two distinct htlcs (they'll have different htlc ids), so the original one lives on independently

Reply:
Ok, clear